typing/typecore.ml

```diff
@@ -6044,7 +6022,8 @@ and type_construct env (expected_mode : expected_mode) loc lid sarg
     end;
   (* NOTE: shouldn't we call "re" on this final expression? -- AF *)
   { texp with
-    exp_desc = Texp_construct(lid, constr, args) }
+    exp_desc = Texp_construct(lid, constr, args,
+      Value_mode.regional_to_global_alloc expected_mode.mode) }
```

Comment:
As with the `Texp_record` case it would probably be better for this mode to be optional so that the unboxed case makes sense. That would also allow to leave it as `None` for immediate constructors.